### PR TITLE
Holiday Stops : shared library (mainly SalesForce integration)

### PR DESCRIPTION
**Pull request by twrichards**
Added new shared library `holiday-stops` (in `lib`) for use by handlers: `holiday-stops-api` (_insert PR reference here_) and `holiday-stops-processor` (https://github.com/guardian/support-service-lambdas/pull/289).

Largely for interacting with new SalesForce objects `Holiday Stop Request` and `Holiday Stop Request Actioned Zuora Ref` (which have a _master-detail_ relationship)

Also provides simple `ActionCalculator` for calculating the dates of issues impacted by holiday stop, this can be used both to present client-side and also compare the count against actioned count so far to determine what still needs to be processed.

### CODE REVIEW

File: lib/holiday-stops/src/main/scala/com/gu/holiday_stops/ActionCalculator.scala [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+import scala.annotation.tailrec
+
+object ActionCalculator {
+
+  def productNameToDayOfWeek(productName: ProductName): Int = productName.value match {
+    case s if s.startsWith("Guardian Weekly") => DateTimeConstants.THURSDAY
+  }
+
+  def calculateActionDatesForGivenHolidayStopRequest(hsr: HolidayStopRequest): List[LocalDate] = {
```
Comment (kelvin-chappell) [inline lib/holiday-stops/src/main/scala/com/gu/holiday_stops/ActionCalculator.scala]: Not sure whether this name is clear.  If list is issue dates to be credited, perhaps could be renamed to something more specific like `publicationDatesToBeStopped`.

File: lib/holiday-stops/src/main/scala/com/gu/holiday_stops/ActionCalculator.scala [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+import com.gu.salesforce.holiday_stops.SalesforceHolidayStopRequest.{HolidayStopRequest, ProductName}
+import org.joda.time.{DateTimeConstants, LocalDate}
+
+import scala.annotation.tailrec
+
+object ActionCalculator {
+
+  def productNameToDayOfWeek(productName: ProductName): Int = productName.value match {
+    case s if s.startsWith("Guardian Weekly") => DateTimeConstants.THURSDAY
```
Comment (kelvin-chappell) [inline lib/holiday-stops/src/main/scala/com/gu/holiday_stops/ActionCalculator.scala]: If this is publication date, I think it's meant to be Friday.

File: lib/holiday-stops/src/main/scala/com/gu/salesforce/holiday_stops/SalesforceHolidayStopRequestActionedZuoraRef.scala [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+import com.gu.util.Logging
+import com.gu.util.resthttp.RestOp._
+import com.gu.util.resthttp.RestRequestMaker._
+import com.gu.util.resthttp.Types.ClientFailableOp
+import com.gu.util.resthttp.{HttpOp, RestRequestMaker}
+import play.api.libs.json.{JsValue, Json}
+
+object SalesforceHolidayStopRequestActionedZuoraRef extends Logging {
+
+  private val sfApiBaseUrl = "/services/data/v29.0"
```
Comment (kelvin-chappell) [inline lib/holiday-stops/src/main/scala/com/gu/salesforce/holiday_stops/SalesforceHolidayStopRequestActionedZuoraRef.scala]: This has already been defined earlier in another place.

File: lib/holiday-stops/src/main/scala/com/gu/salesforce/holiday_stops/SalesforceHolidayStopRequest.scala [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+  }
+
+  object LookupByIdentityId {
+
+    def apply(sfGet: HttpOp[RestRequestMaker.GetRequestWithParams, JsValue]): String => ClientFailableOp[List[HolidayStopRequest]] =
+      sfGet.setupRequest(toRequest).parse[HolidayStopRequestSearchQueryResponse].map(_.records).runRequest
+
+    def toRequest(identityId: String) = {
+      val soqlQuery = HOLIDAY_STOP_REQUEST_SOQL_PREFIX +
```
Comment (paulbrown1982) [inline lib/holiday-stops/src/main/scala/com/gu/salesforce/holiday_stops/SalesforceHolidayStopRequest.scala]: What happens if the Identity ID is blank?

File: lib/holiday-stops/src/main/scala/com/gu/salesforce/holiday_stops/SalesforceHolidayStopRequest.scala [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+import org.joda.time.LocalDate
+import org.joda.time.format.DateTimeFormat
+import play.api.libs.json.{Format, JsResult, JsString, JsValue, Json}
+
+object SalesforceHolidayStopRequest extends Logging {
+
+  val SALESFORCE_DATE_FORMAT = "yyyy-MM-dd"
+
+  private val sfApiBaseUrl = "/services/data/v29.0"
+  private val soqlQueryBaseUrl = RelativePath(sfApiBaseUrl + "/query/")
```
[the remark this answers is not in the document]
Comment (twrichards) [reply lib/holiday-stops/src/main/scala/com/gu/salesforce/holiday_stops/SalesforceHolidayStopRequest.scala]: cleaned up in https://github.com/guardian/support-service-lambdas/pull/294/commits/5761a77d8d79e282495be630c24a9e5fd9b28413

File: lib/holiday-stops/src/main/scala/com/gu/salesforce/holiday_stops/SalesforceHolidayStopRequestActionedZuoraRef.scala (hunk 3 above)
Comment (twrichards) [reply lib/holiday-stops/src/main/scala/com/gu/salesforce/holiday_stops/SalesforceHolidayStopRequestActionedZuoraRef.scala]: https://github.com/guardian/support-service-lambdas/pull/294/commits/5761a77d8d79e282495be630c24a9e5fd9b28413

File: lib/holiday-stops/src/main/scala/com/gu/salesforce/holiday_stops/SalesforceHolidayStopRequest.scala (hunk 4 above)
Comment (twrichards) [reply lib/holiday-stops/src/main/scala/com/gu/salesforce/holiday_stops/SalesforceHolidayStopRequest.scala]: no results come back - which is fine

File: lib/holiday-stops/src/main/scala/com/gu/holiday_stops/ActionCalculator.scala [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    @tailrec
+    def processNextDay(listSoFar: List[LocalDate], date: LocalDate): List[LocalDate] =
+      if (date == stopRecursionAt)
+        listSoFar
+      else if (date.getDayOfWeek == dayOfWeekForProduct)
+        processNextDay(date :: listSoFar, date.plusDays(1))
+      else
+        processNextDay(listSoFar, date.plusDays(1))
+
+    processNextDay(List(), hsr.Start_Date__c.value).reverse
```
[the remark this answers is not in the document]
Comment (kelvin-chappell) [reply lib/holiday-stops/src/main/scala/com/gu/holiday_stops/ActionCalculator.scala]: Actually may not be possible.  Ignore this.

[the remark this answers is not in the document]
Comment (kelvin-chappell) [reply lib/holiday-stops/src/main/scala/com/gu/holiday_stops/ActionCalculator.scala]: Was thinking of something like (using java time):
```
def applicableDates(from:LocalDate, to: LocalDate, p: LocalDate => Boolean): Seq[LocalDate] = {
  (0l to ChronoUnit.DAYS.between(from, to)).foldLeft(Seq.empty[LocalDate]){(acc, i) =>
  val d = from.plusDays(i)
  if(p(d)) acc :+ d
  else acc
}}
```
eg. 
```
applicableDates(LocalDate.now, LocalDate.now.plusDays(21), {_.getDayOfWeek == DayOfWeek.FRIDAY})
```

File: build.sbt [hunk 7]
```diff
@@ -88,6 +88,19 @@ lazy val salesforce = all(project in file("lib/salesforce"))
     libraryDependencies ++= Seq(okhttp3, logging, scalaz, playJson, scalatest)
   )
 
+lazy val `holiday-stops` = all(project in file("lib/holiday-stops"))
+  .dependsOn(
+    salesforce,
+    restHttp,
+    handler,// % "test->test" TODO make this dep only in test - SF client shouldn't depends on ApiGateway
```
[the remark this answers is not in the document]
Comment (twrichards) [reply build.sbt]: cleaned up this list in https://github.com/guardian/support-service-lambdas/pull/294/commits/1273792bc484171a26d0c0a66ceb5fccbbe4e38d

**Pull request merged**